pkg/ring/basic_lifecycler_delegates.go

```diff
@@ -0,0 +1,98 @@
+package ring
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+)
+
+type LeaveOnStoppingDelegate struct {
+	next   BasicLifecyclerDelegate
+	logger log.Logger
+}
+
+func NewLeaveOnStoppingDelegate(next BasicLifecyclerDelegate, logger log.Logger) *LeaveOnStoppingDelegate {
+	return &LeaveOnStoppingDelegate{
+		next:   next,
+		logger: logger,
+	}
+}
+
+func (d *LeaveOnStoppingDelegate) OnRingInstanceRegister(lifecycler *BasicLifecycler, ringDesc Desc, instanceExists bool, instanceID string, instanceDesc IngesterDesc) (IngesterState, Tokens) {
+	return d.next.OnRingInstanceRegister(lifecycler, ringDesc, instanceExists, instanceID, instanceDesc)
+}
+
+func (d *LeaveOnStoppingDelegate) OnRingInstanceTokens(lifecycler *BasicLifecycler, tokens Tokens) {
+	d.next.OnRingInstanceTokens(lifecycler, tokens)
+}
+
+func (d *LeaveOnStoppingDelegate) OnRingInstanceStopping(lifecycler *BasicLifecycler) {
+	if err := lifecycler.changeState(context.Background(), LEAVING); err != nil {
+		level.Error(d.logger).Log("msg", "failed to change instance state to LEAVING in the ring", "err", err)
+	}
+
+	d.next.OnRingInstanceStopping(lifecycler)
+}
+
+type TokensPersistencyDelegate struct {
+	next       BasicLifecyclerDelegate
+	logger     log.Logger
+	tokensPath string
+	loadState  IngesterState
+}
+
+func NewTokensPersistencyDelegate(path string, state IngesterState, next BasicLifecyclerDelegate, logger log.Logger) *TokensPersistencyDelegate {
+	return &TokensPersistencyDelegate{
+		next:       next,
+		logger:     logger,
+		tokensPath: path,
+		loadState:  state,
+	}
+}
+
+func (d *TokensPersistencyDelegate) OnRingInstanceRegister(lifecycler *BasicLifecycler, ringDesc Desc, instanceExists bool, instanceID string, instanceDesc IngesterDesc) (IngesterState, Tokens) {
+	// Skip if no path has been configured.
+	if d.tokensPath == "" {
+		level.Info(d.logger).Log("msg", "not loading tokens from file, tokens file path is empty")
+		return d.next.OnRingInstanceRegister(lifecycler, ringDesc, instanceExists, instanceID, instanceDesc)
+	}
+
+	// Do not load tokens from disk if the instance is already in the ring.
+	if instanceExists {
```

Comment:
It may exist, but not have tokens.

Reply:
Right! Fixed and covered with a test. The previous implementation was inherited by `ring.Lifecycler` (which doesn't handle this case) but I think what you suggest is better.